src/main/java/org/commcare/cases/entity/EntityStringFilterer.java

```diff
@@ -0,0 +1,57 @@
+package org.commcare.cases.entity;
+
+import org.commcare.modern.util.Pair;
+import org.commcare.util.EntitySortUtil;
+import org.javarosa.core.model.instance.TreeReference;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+
+/**
+ * Filter entity list via all string-representable entity fields
+ */
+public class EntityStringFilterer {
+    private final String[] searchTerms;
+    private final ArrayList<Pair<Integer, Integer>> matchScores = new ArrayList<>();
+
+    private final NodeEntityFactory nodeFactory;
+    protected final List<Entity<TreeReference>> matchList;
+    protected final List<Entity<TreeReference>> fullEntityList;
+    private final boolean isFuzzySearchEnabled;
+
+    public EntityStringFilterer(String[] searchTerms,
+            NodeEntityFactory nodeFactory,
+            List<Entity<TreeReference>> fullEntityList,
+            boolean isFuzzySearchEnabled) {
+        this.isFuzzySearchEnabled = isFuzzySearchEnabled;
+        this.matchList = new ArrayList<>();
+        this.nodeFactory = nodeFactory;
+        this.fullEntityList = fullEntityList;
+        this.searchTerms = searchTerms;
+        if (searchTerms == null || searchTerms.length == 0) {
+            matchList.addAll(fullEntityList);
+        }
+    }
+
+    public List<Entity<TreeReference>> buildMatchList() {
+
+        while (!nodeFactory.isEntitySetReady()) {
+            try {
+                Thread.sleep(100);
```

Comment:
I know you just copied this but I checked the logic here and there's no multithreading so I don't see why this is necessary.

I do see that mobile implements an async version of the factory. Perhaps a task for future refactor to clean this up.

Reply:
agree this is not necessary in FP context and  is there to only ensure the mobile compatibility with this class although mobile implements it's own filter class today. I will be happy to remove this code from here but think we will need to explicitly mention that this class should only be used for CLI/FP. Maybe by moving it to `src/cli` ?